File: src/test/baekjoon/python/greedy/ClassroomAssignment.py
import heapq

N = int(input())
times = []
for _ in range(N):
    times.append(list(map(int, input().split())))

times.sort()

answer = []
heapq.heappush(answer, times[0][1])

for i in range(1, N):
    if times[i][0] >= answer[0]:
        heapq.heappop(answer)

    heapq.heappush(answer, times[i][1])

print(len(answer))

File: src/test/baekjoon/java/greedy/ClassroomAssignment.java
package test.baekjoon.java.greedy;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Time {
    int start;
    int end;

    public Time(int start, int end) {
        this.start = start;
        this.end = end;
    }
}

public class ClassroomAssignment {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        List<Time> times = new ArrayList<>();
        StringTokenizer st;
        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            int start = Integer.parseInt(st.nextToken());
            int end = Integer.parseInt(st.nextToken());

            times.add(new Time(start, end));
        }

        times.sort(Comparator.comparingInt(o -> o.start));

        PriorityQueue<Integer> answer = new PriorityQueue<>();
        answer.add(times.get(0).end);

        for (int i = 1; i < N; i++) {
            if (times.get(i).start >= answer.peek()) {
                answer.poll();
            }

            answer.add(times.get(i).end);
        }

        System.out.println(answer.size());
    }
}
